src/pages/snake.astro

```diff
@@ -0,0 +1,222 @@
+    } while (snake.body.some((segment) => segment.x === newFoodX && segment.y === newFoodY));
+
+    food.x = newFoodX;
+    food.y = newFoodY;
+  }
+
+  window.addEventListener('keydown', (event) => {
+    event.preventDefault();
+    if (gameOver && event.key === ' ') {
+      resetGame();
+      return;
+    }
+    switch (event.key) {
+      case 'ArrowUp':
+      case 'w':
+      case 'k':
+        if (snake.dy !== 1) {
+          snake.dx = 0;
+          snake.dy = -1;
+        }
+        break;
+      case 'ArrowDown':
+      case 's':
+      case 'j':
+        if (snake.dy !== -1) {
+          snake.dx = 0;
+          snake.dy = 1;
+        }
+        break;
+      case 'ArrowLeft':
+      case 'a':
+      case 'h':
+        if (snake.dx !== 1) {
+          snake.dx = -1;
+          snake.dy = 0;
+        }
+        break;
+      case 'ArrowRight':
+      case 'd':
+      case 'l':
+        if (snake.dx !== -1) {
+          snake.dx = 1;
+          snake.dy = 0;
+        }
+        break;
+    }
+  });
+
+  window.addEventListener('touchstart', handleTouchStart, false);
+  window.addEventListener('touchmove', handleTouchMove, false);
+
+  let xDown = null;
+  let yDown = null;
+
+  function handleTouchStart(event) {
+    if (gameOver) resetGame();
+
+    const firstTouch = event.touches[0];
+    xDown = firstTouch.clientX;
+    yDown = firstTouch.clientY;
+  }
+
+  function handleTouchMove(event) {
+    if (!xDown || !yDown) return;
+
+    const xUp = event.touches[0].clientX;
+    const yUp = event.touches[0].clientY;
+
+    const xDiff = xDown - xUp;
+    const yDiff = yDown - yUp;
+
+    if (Math.abs(xDiff) > Math.abs(yDiff)) {
+      if (xDiff > 0) {
+        if (snake.dx !== 1) {
+          snake.dx = -1;
+          snake.dy = 0;
+        }
+      } else {
+        if (snake.dx !== -1) {
+          snake.dx = 1;
+          snake.dy = 0;
+        }
+      }
+    } else {
+      if (yDiff > 0) {
+        if (snake.dy !== 1) {
+          snake.dx = 0;
+          snake.dy = -1;
+        }
+      } else {
+        if (snake.dy !== -1) {
+          snake.dx = 0;
+          snake.dy = 1;
+        }
+      }
+    }
+    xDown = null;
+    yDown = null;
+  }
+
+  function resetGame() {
+    snake.body = [{ x: 10, y: 10 }];
+    snake.dx = 1;
+    snake.dy = 0;
+    food.x = Math.floor(Math.random() * 20);
+    food.y = Math.floor(Math.random() * 20);
+    gameOver = false;
+  }
+
+  async function main() {
+    const canvas = document.createElement('canvas');
+    const ctx = canvas.getContext('2d');
+    canvas.width = window.innerWidth;
+    canvas.height = window.innerHeight;
+    document.body.appendChild(canvas);
+
+    snake.body.push({ x: 10, y: 10 });
+
+    setInterval(() => {
+      moveSnake(ctx);
+      draw(ctx);
+    }, 1000 / FRAME_RATE);
+  }
+
+  main();
+</script>
+
+<style is:inline>
+  body {
+    margin: 0;
+    overflow: hidden;
+    padding: 0;
+    border: 0;
+  }
```

Comment:
Our friends at Firefox Mobile can still do the "pull to refresh" action. That can be fixed by adding `touch-action: none` to the `body` selector.